tests/python/unittest/test_halide_text_format.py

```diff
@@ -0,0 +1,163 @@
+    tvm_c = tvm.ndarray.array(numpy.zeros((_n, _m), dtype = 'float32'))
+    func(_n, _m, tvm_a, tvm_b, tvm_c)
+    numpy.testing.assert_allclose(tvm_c.asnumpy(), c_python, rtol = 1e-5)
+    #print(func)
+
+def fanout(n, a, b):
+    three = 3.0
+    for i in range(n - 3):
+        sigma = 0.0
+        for j in range(3):
+            sigma = sigma + a[i + j]
+        sigma = sigma / three
+        b[i] = sigma
+
+def test_fanout():
+    n = tvm.var('n')
+    a = tvm.placeholder((n, ), name = 'a')
+    b = tvm.placeholder((n-3, ), name = 'b')
+    ir = frontend.parse(fanout, [n, a, b])
+    #print(ir)
+    assert isinstance(ir, tvm.stmt.Realize)
+    assert ir.bounds[0].min.value == 0
+    assert ir.bounds[0].extent.value == 1
+    assert ir.func.name == 'sigma'
+    rbody = ir.body
+    #Check for i in (0, n-3)
+    assert isinstance(rbody, tvm.stmt.For)
+    assert rbody.loop_var.name == 'i'
+    assert rbody.min.value == 0
+    assert tvm.ir_pass.Equal(rbody.extent, n - 3)
+    #Check i loop body
+    ibody = rbody.body
+    assert isinstance(ibody.first, tvm.stmt.Provide)
+    assert ibody.first.func.name == 'sigma'
+    assert len(ibody.first.args) == 1
+    assert ibody.first.args[0].value == 0
+    #Check fanout loop
+    jloop = ibody.rest.first
+    assert jloop.loop_var.name == 'j'
+    assert jloop.min.value == 0
+    assert jloop.extent.value == 3
+    jbody = jloop.body
+    assert isinstance(jbody, tvm.stmt.Provide)
+    assert len(jbody.args) == 1
+    assert jbody.args[0].value == 0
+    assert jbody.func.name == 'sigma'
+    assert isinstance(jbody.value, tvm.expr.Add)
+    value = jbody.value
+    assert isinstance(value.a, tvm.expr.Call)
+    assert value.a.name == 'sigma'
+    assert len(value.a.args) == 1
+    assert value.a.args[0].value == 0
+    assert value.b.name == 'a'
+    assert len(value.b.args) == 1
+    assert tvm.ir_pass.Equal(value.b.args[0], rbody.loop_var + jloop.loop_var)
+    divide= ibody.rest.rest.first
+    assert isinstance(divide, tvm.stmt.Provide)
+    assert len(divide.args) == 1
+    assert divide.args[0].value == 0
+    value = divide.value
+    assert isinstance(value, tvm.expr.Mul)
+    assert value.a.name == 'sigma'
+    assert len(value.a.args) == 1
+    assert value.a.args[0].value == 0
+    assert abs(value.b.value - (1 / 3.0)) < 1e-5
+    write = ibody.rest.rest.rest
+    assert isinstance(write, tvm.stmt.Provide)
+    assert write.func.name == 'b'
+    assert write.value.name == 'sigma'
+    assert len(write.value.args) == 1
+    assert write.value.args[0].value == 0
+    #print(write)
+
+
+def failure():
+    for i in range(1, 100):
+        i = 0
+
+def test_failure():
+    try:
+        frontend.parse(failure, [])
+    except AssertionError:
+        _, _, tb = sys.exc_info()
+        _, _, func, text = traceback.extract_tb(tb)[-1]
+        assert func == 'visit_Name'
+        assert text == 'assert node.id not in self.loops_above'
+
+@tvm.contrib.pyfrontend.py_frontend
```

Comment:
IMO, we may put this frontend in tvm/python instead of contrib.

Reply:
@tqchen

Reply:
Maybe it is a good idea, but Tianqi told me that we should put immature functionality in `contrib` first.